src/rebar_agent.erl

```diff
@@ -179,8 +237,27 @@ refresh_state(RState, _Dir) ->
 %% @private takes a list of modules and reloads them
 -spec reload_modules([module()]) -> term().
 reload_modules([]) -> noop;
-reload_modules(Modules) ->
-        reload_modules(Modules, erlang:function_exported(code, prepare_loading, 1)).
+reload_modules(Modules0) ->
+    Modules = [M || M <- Modules0, is_changed(M)],
+    reload_modules(Modules, erlang:function_exported(code, prepare_loading, 1)).
+
+%% @spec is_changed(atom()) -> boolean()
+%% @doc true if the loaded module is a beam with a vsn attribute
+%%      and does not match the on-disk beam file, returns false otherwise.
+is_changed(M) ->
+    try
+        module_vsn(M:module_info()) =/= module_vsn(code:get_object_code(M))
```

Comment:
This check will fail when someone wants to hot-reload code they are testing but explicitly set a `-vsn` module attribute such as when trying to do hot-code loading with relups. In that case, many code modifications can take place without `-vsn(...)` ever changing. This breaks things for such use cases.

I think it might be safer to just not do filtering here.

Reply:
Yes, this is really a problem, but if you don't use filtering, it will cause a lot of modules to be reloaded. Even if the module has not changed. 
At the beginning I want to change to the last modified timestamp, but it have not saved the last modification time.

Reply:
Right. That's why I think we could work around that by auto-filtering when using MD5s (which is safe to do), and not filtering with the `vsn` is _not_ an MD5 -- in such a case, it is probably better to always reload.